File: tests/ApiTests.java
package com.cloudapplication;

import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.web.client.TestRestTemplate;
import org.springframework.http.*;
import org.springframework.test.context.DynamicPropertyRegistry;
import org.springframework.test.context.DynamicPropertySource;

import java.util.Map;

import static org.assertj.core.api.Assertions.assertThat;

@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
class ApiTests {

    @Autowired
    private TestRestTemplate restTemplate;

    // Configura dinamicamente a conexão com o PostgreSQL para os testes, 
    // equivalendo ao ajuste da string de conexão no WebHost.
    @DynamicPropertySource
    static void configureProperties(DynamicPropertyRegistry registry) {
        String dbUrl = System.getenv("DATABASE_URL");
        if (dbUrl == null || dbUrl.isEmpty()) {
            dbUrl = "jdbc:postgresql://localhost:5432/orders";
        } else {
            // Nota: No Spring, URLs do JDBC geralmente começam com jdbc:postgresql://
            // Dependendo de como a string de ambiente vem, pode ser necessário ajustar o prefixo.
        }

        registry.add("spring.datasource.url", () -> dbUrl);
        registry.add("spring.datasource.username", () -> "postgres");
        registry.add("spring.datasource.password", () -> "postgres");
        registry.add("spring.jpa.hibernate.ddl-auto", () -> "create-drop"); // Equivalente ao EnsureCreated()
    }

    @Test
    void testHealth() {
        ResponseEntity<Map> response = restTemplate.getForEntity("/health", Map.class);
        
        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.OK);
        
        Map<String, String> json = response.getBody();
        assertThat(json).isNotNull();
        assertThat(json).containsKey("status");
        assertThat(json).containsKey("database");
    }

    @Test
    void testCreateOrder() {
        Map<String, Object> newOrder = Map.of("customer", "Maria");
        
        ResponseEntity<Order> response = restTemplate.postForEntity("/orders", newOrder, Order.class);
        
        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.CREATED);
        
        Order order = response.getBody();
        assertThat(order).isNotNull();
        assertThat(order.getCustomer()).isEqualTo("Maria");
        assertThat(order.getStatus()).isEqualTo("Created");
        assertThat(order.getId()).isNotNull();
    }
}

File: tests/ApiTests.php
namespace CloudApplication;

use PHPUnit\Framework\TestCase;
use GuzzleHttp\Client;
use GuzzleHttp\Exception\RequestException;

class ApiTest extends TestCase
{
    private Client $client;
    private string $baseUrl;

    protected function setUp(): void
    {
        parent::setUp();
        
        // Define a URL base da aplicação para os testes
        $this->baseUrl = $_ENV['APP_URL'] ?? 'http://localhost:8080';

        $this->client = new Client([
            'base_uri' => $this->baseUrl,
            'http_errors' => false, // Permite capturar códigos de status diferentes de 2xx
            'headers' => [
                'Accept' => 'application/json',
                'Content-Type' => 'application/json',
            ],
        ]);
    }

    /**
     * Equivalente ao ajuste de propriedades de conexão e ddl-auto.
     * Em PHP, a configuração do banco e migrações geralmente é feita via variáveis de ambiente 
     * no arquivo phpunit.xml ou bootstrap do framework (ex: Laravel/Symfony).
     */
    public static function setUpBeforeClass(): void
    {
        parent::setUpBeforeClass();

        $dbUrl = getenv('DATABASE_URL');
        if (empty($dbUrl)) {
            $dbUrl = 'pgsql://postgres:postgres@localhost:5432/orders';
        }

        // Exemplo de definição em tempo de execução para o ambiente de testes
        $_ENV['DB_CONNECTION'] = 'pgsql';
        $_ENV['DB_URL'] = $dbUrl;
    }

    public function testHealth(): void
    {
        $response = $this->client->get('/health');

        $this->assertEquals(200, $response->getStatusCode());

        $json = json_decode($response->getBody()->getContents(), true);

        $this->assertNotNull($json);
        $this->assertArrayHasKey('status', $json);
        $this->assertArrayHasKey('database', $json);
    }

    public function testCreateOrder(): void
    {
        $newOrder = [
            'customer' => 'Maria'
        ];

        $response = $this->client->post('/orders', [
            'json' => $newOrder
        ]);

        $this->assertEquals(201, $response->getStatusCode());

        $order = json_decode($response->getBody()->getContents(), true);

        $this->assertNotNull($order);
        $this->assertEquals('Maria', $order['customer'] ?? null);
        $this->assertEquals('Created', $order['status'] ?? null);
        $this->assertNotNull($order['id'] ?? null);
    }
}